ci: Add "macOS 12 native arm64" task

hebasto opened the pull request:
The release binaries for Apple Silicon macOS (`arm64-apple-darwin`) have been available since v23.0.

It seems reasonable to add a task for such a platform to CI. Also it would prevent regressions like bitcoin/bitcoin#24958.

hebasto committed:
ci, refactor: Add `MACOS_NATIVE_TASK_TEMPLATE`

A native macOS task does not aware of Linux container settings, and it
does not use the `depends_built_cache`.

hebasto committed:
ci: Add "macOS 12 native Apple Silicon" task

hebasto committed:
ci: Improve naming related to "macOS 12 native x86_64" task

MarcoFalke commented:
Would this actually prevent #24958? Wouldn't it silently "fail" green?

hebasto commented:
> Would this actually prevent #24958? Wouldn't it silently "fail" green?

https://github.com/bitcoin/bitcoin/blob/f2e07bdbac58c2cd38cd5f668847e95cde3eeefd/ci/test/00_setup_env_mac_native_arm64.sh#L11

It will fail due to https://github.com/bitcoin/bitcoin/blob/f7a1e676d536115aba649672ba1e68755a1c90dc/configure.ac#L1751-L1757

Zero-1729 commented:
Concept ACK

MarcoFalke commented:
Not sure if it makes sense to run this check on every commit. There is some probability of breaking that is acceptable and unavoidable. I think we should focus on checks that are likely to catch errors often enough to be a net-positive considering the maintenance overhead and CPU overhead of the added CI task.

So if https://github.com/bitcoin/bitcoin/pull/24958 is the only known bug this prevents, I tend toward NACK, even though the code looks good and correct.

I expect that there are ppl running M1 locally, but if not, this could also be implemented as a nightly task to run outside this repo.

promag commented on ci/test/04_install.sh:
```diff
@@ -10,10 +10,10 @@ if [[ $QEMU_USER_CMD == qemu-s390* ]]; then
   sudo -H pip3 install --upgrade pip
   # shellcheck disable=SC2086
-  IN_GETOPT_BIN="/usr/local/opt/gnu-getopt/bin/getopt" ${CI_RETRY_EXE} pip3 install --user $PIP_PACKAGES
+  IN_GETOPT_BIN="$(brew --prefix gnu-getopt)/bin/getopt" ${CI_RETRY_EXE} pip3 install --user $PIP_PACKAGES
```
Nice one 🤓

promag commented on .cirrus.yml:
```diff
@@ -306,18 +314,28 @@ task:
+    FILE_ENV: "./ci/test/00_setup_env_mac_native_x86_64.sh"
+
+task:
+  name: 'macOS 12 native Apple Silicon [gui, system sqlite] [no depends]'
```
nit, s/Apple Silicon/arm64

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/25160#issuecomment-1135487448).

promag reviewed (commented):
Concept ACK

hebasto commented:
> CPU overhead of the added CI task

I don't think the M1 CPU time is shared with any other task.

MarcoFalke commented:
According to https://cirrus-ci.org/faq/#are-there-any-limits  there is a limit of "12.0 CPUs macOS VM (1 VM)", so it will be shared with the other macos instance?

hebasto commented:
> According to https://cirrus-ci.org/faq/#are-there-any-limits there is a limit of "12.0 CPUs macOS VM (1 VM)", so it will be shared with the other macos instance?

It looks like docs are a bit inconsistent after introducing Apple Silicon support.

They are correct about [`x86_64`](https://cirrus-ci.com/task/5473695194939392):
![image](https://user-images.githubusercontent.com/32963518/169489581-f43fbde2-7cfe-49e3-8c51-4e6a69987837.png)

But M1 limits [differ](https://cirrus-ci.com/task/6599595101782016):
![image](https://user-images.githubusercontent.com/32963518/169489776-222cf449-696b-4e7c-9abb-63ef4988fffc.png)

MarcoFalke commented:
Ok, looks like it won't affect the limits for now. Though, I am still wondering how many bugs this will catch that are not found by other tasks.

hebasto commented:
Updated f2e07bdbac58c2cd38cd5f668847e95cde3eeefd -> 84cfbc7329e92191e391ba5bb1991108229bbb51 ([pr25160.02](https://github.com/hebasto/bitcoin/commits/pr25160.02) -> [pr25160.03](https://github.com/hebasto/bitcoin/commits/pr25160.03), [diff](https://github.com/hebasto/bitcoin/compare/pr25160.02..pr25160.03)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/25160#discussion_r876177791)

> I am still wondering how many bugs this will catch that are not found by other tasks.

I hope every bug on the macOS `arm64` platform, which has a steadily growing number of users, will be caught before release.

Zero-1729 reviewed (commented):
crACK 84cfbc7329e92191e391ba5bb1991108229bbb51

MarcoFalke commented:
> I hope every bug on the macOS arm64 platform, which has a steadily growing number of users, will be caught before release.

It only caught one bug for now, which doesn't seem enough to justify the maintenance burden? Note that in this repo, *bsd, s390x, various linux distros, valgrind, ... aren't tested either.

MarcoFalke commented:
I guess the 1st+3rd commit could still make sense. Then I could add it to my btc_nightly CI run, if you want.

MarcoFalke commented:
if amd64 is unsupported, we should probably remove it and just switch to arm64?
